Escape hatch

ImJeremyHe opened the pull request:
Closes #300 
The test for escape hatch is not ready.

### This PR:
- Adds the escape hatch in the batch poster. The escape hatch is a function being called periodically, and it works as following:
    - if hotshot was dwon, check if hotshot is live
    - if hotshot was up, check if hotshot is down. If there is also a hotshot transaction waiting for being finalized, then check hotshot liveness of its corresponding L1 height. If down, remove the pending queue and submitted positions and set a value indicating messages skipping espresso validation.
    - if hotshot is down, and escape hatch is on. Then espresso validation will be circumvented
    - if hotshot is down, and escape hatch is off. Then batch poster will submit transactions to hotshot when it goes back live
- Refactors a bit to make the workflow more comprehensible.
- Fixes submitting positions to the pending queue
- Add a specific escape hatch for it

### This PR does not:
test the escape hatch totally. Now we don't have a good way to see if any hotshot transactions occur when hotshot is down and escape hatch is on. Or we can think of any other way to test the escape hatch fully. However, this PR has already included so many stuff. Getting it merged first can reduce conflicts

### Key places to review:
batch_poster.go
transaction_streamer.go

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->

ImJeremyHe committed:
Create SwitchSequencer

ImJeremyHe committed:
Update the espresso-sequencer-go

ImJeremyHe committed:
Use light client as the trigger

ImJeremyHe committed:
Add the switch test

ImJeremyHe committed:
Make the linter happy

ImJeremyHe committed:
Add the switch test to the e2e ci

ImJeremyHe committed:
Increase the limit

ImJeremyHe committed:
Merge pull request #110 from EspressoSystems/jh/escape-hatch

Add SwitchSequencer

ImJeremyHe committed:
Clean the log with the log helper

nomaxg committed:
more mock code

nomaxg committed:
merge

ImJeremyHe committed:
Fix the e2e test

nomaxg committed:
serialize proof variables correctly

nomaxg committed:
update espresso-go version

nomaxg committed:
wasm compatibility

nomaxg committed:
remove stale file, lint fix

nomaxg committed:
merge e2e fixes

nomaxg committed:
update to recent espresso go version

nomaxg committed:
refactor block merkle verification

nomaxg committed:
work on E2E

nomaxg committed:
progress on merkle tree verification

nomaxg committed:
add test vectors, fetch proofs at the correct indicies

nomaxg committed:
cleanup, remove outfiles, fetch merkle proof at the correct justified height

nomaxg committed:
add logic to compare circuit field

nomaxg committed:
check that circuit representation equals u256 representation in the test

nomaxg committed:
send byte slice to wasm instead of byte array which lives on the stack

nomaxg committed:
cleanup, remove debug statements, add documentation

nomaxg committed:
remove error logs

nomaxg committed:
fix build

nomaxg committed:
lint fix

nomaxg committed:
progress on e2e tests, comment out vid stuff for now because of wasm io issues

nomaxg committed:
update espresso-sequencer-go tag, lint ci fix

nomaxg committed:
cleanup, add issues for unresolved problems

nomaxg committed:
remove block that spams transactions

nomaxg committed:
comment out unused txns

nomaxg committed:
add test for advz equivalence and use a smaller srs

ImJeremyHe committed:
Allow dead code in validation

ImJeremyHe committed:
Specify foundry version

ImJeremyHe committed:
Add random_get stub in wasi_snapshot_preview1

ImJeremyHe committed:
Add stubs required in wasi_snapshot_preview1

sveitser committed:
Fix `make lint`

sveitser committed:
Enable prover service to talk to L1 on host

sveitser committed:
Add builder timeout

This allows the sequencer to make progress past block 1.

These env vars have been changed in espresso-sequencer.

sveitser committed:
Add missing env var, more logging in e2e test

sveitser committed:
Revert "Fix `make lint`"

This reverts commit 8be940c2ce941e7d65cd44400d4e161bb7c70cc7.

sveitser committed:
Fix go linting

ImJeremyHe committed:
Parse espresso messages

ImJeremyHe committed:
Add timeout limit

ImJeremyHe committed:
Fix the e2e test

ImJeremyHe committed:
Add timeout limit

ImJeremyHe committed:
Run the e2e test and switch test in 1 thread

ImJeremyHe committed:
Update the e2e workflow

ImJeremyHe committed:
Increase time limit

nomaxg committed:
Merge pull request #111 from EspressoSystems/mock-validation

Light Client States Updates

ImJeremyHe committed:
Specify the foundryup version in Dockerfile

ImJeremyHe committed:
Add sock_accept stub

ImJeremyHe committed:
Increase timeout limit

ImJeremyHe committed:
Merge pull request #118 from EspressoSystems/jh/images

Specify the foundryup version in Dockerfile

ImJeremyHe committed:
Merge branch 'master' of github.com:OffchainLabs/nitro into jh/sync

ImJeremyHe committed:
Build successfully

nomaxg committed:
move ark-srs to dev dependencies

nomaxg committed:
remove debug statements

nomaxg committed:
remove out.txt

nomaxg committed:
update lockfile

ImJeremyHe committed:
Fix the machine test

ImJeremyHe committed:
Fix CI

ImJeremyHe committed:
Fix switch test

ImJeremyHe committed:
Fix CI

ImJeremyHe committed:
Increase CI time limie

nomaxg committed:
remove commented out replaces

nomaxg committed:
newline

ImJeremyHe committed:
Use the latest foundry

ImJeremyHe committed:
RESET ME: debug

ImJeremyHe committed:
RESET ME: not run the CI in a container

ImJeremyHe committed:
RESET ME: remove the parallel

sveitser committed:
Update script to run tests locally

- Run the same commands as on CI.
- Add ability to run `cargo +nightly` to dev shell.
- make clean: remove brotli build files.

sveitser committed:
set -parallel=2

nomaxg committed:
Merge pull request #121 from EspressoSystems/nix-validation-fix

Fix VID Validation

ImJeremyHe committed:
Merge remote-tracking branch 'origin/integration' into jh/sync

ImJeremyHe committed:
Fix e2e test

ImJeremyHe committed:
Skip some tests in CI

ImJeremyHe committed:
Skip tests in stylus

sveitser committed:
Remove parallelism restriction

ImJeremyHe committed:
Merge pull request #122 from EspressoSystems/jh/sync

Sync upstream 20240513

ImJeremyHe committed:
Impl escape hatch on validator side

ImJeremyHe committed:
Fix tests

ImJeremyHe committed:
Generate the osp of hotshot liveness

ImJeremyHe committed:
Merge pull request #123 from EspressoSystems/jh/escape

Impl escape hatch on validator side

ImJeremyHe committed:
Fix the fraud proof (#124)

* Update contracts submodule

* Upgrade go client

* Fix the staker challenge tests
---------

Co-authored-by: sveitser <sveitser@gmail.com>

ImJeremyHe committed:
Use espresso dev node

nomaxg committed:
attempt to fix dockerfile

nomaxg committed:
newline

nomaxg committed:
Merge pull request #126 from EspressoSystems/fix-dockerfile

Attempt to fix Espresso dockerfile

ImJeremyHe committed:
WIP

ImJeremyHe committed:
WIP

nomaxg committed:
download latest espresso module root

nomaxg committed:
Merge pull request #128 from EspressoSystems/fix-dockerfile

Dockerfile - Download latest espresso module root

nomaxg committed:
incorporate vid common data

ImJeremyHe committed:
Fix the validation pipeline

nomaxg committed:
update go modules, e2e build step

nomaxg committed:
add cbrotli wasm

nomaxg committed:
ci fix attempt

nomaxg committed:
dont build all test deps for e2e test

nomaxg committed:
lint fixes

ImJeremyHe committed:
Use the self-hosted runner and install cbindgen in CI

ImJeremyHe committed:
Add time limit

ImJeremyHe committed:
Remove parallel in building

ImJeremyHe committed:
Fix the switch test

ImJeremyHe committed:
Decrease the waiting hotshot blocks

ImJeremyHe committed:
Use the ubuntu

ImJeremyHe committed:
Specify the foundry version

ImJeremyHe committed:
Enable -j

nomaxg committed:
Merge pull request #127 from EspressoSystems/jh/dev-node

Utilize espresso-dev-node in e2e test and fix the validation pipeline

nomaxg committed:
update foundry version

nomaxg committed:
Merge branch 'integration' into foundry-version

nomaxg committed:
Merge pull request #130 from EspressoSystems/foundry-version

Update foundry version

ImJeremyHe committed:
Fix the staker test

nomaxg committed:
module root update

nomaxg committed:
Merge pull request #133 from EspressoSystems/update-module-root

module root update

nomaxg committed:
increase validation timeout for switch test

ImJeremyHe committed:
Merge pull request #132 from EspressoSystems/jh/fraud

Fix the staker test

zacshowa committed:
pin espresso-sequencer-go to new release v0.0.20

zacshowa committed:
Commit go.sum from running go mod tidy

nomaxg committed:
sequencer info logs

nomaxg committed:
Merge pull request #138 from EspressoSystems/logging

Add sequencer info logs

ImJeremyHe committed:
Merge pull request #137 from zacshowa/update-sequencer-go

pin espresso-sequencer-go to new release v0.0.20

ImJeremyHe committed:
Add escape hatch test in e2e test

ImJeremyHe committed:
Check the reactivate hotshot

ImJeremyHe committed:
Merge pull request #139 from EspressoSystems/jh/e2e

Add escape hatch test in e2e test

zacshowa committed:
Update validator to handle preconditions and panics before entering the espresso STF logic.

zacshowa committed:
Improve readability of conditions in `handleEspressoPreConditions` and update the functions comments.

zacshowa committed:
Further improve readability

zacshowa committed:
Fix lint issue

ImJeremyHe committed:
Update the nitro testnode

ImJeremyHe committed:
Merge branch 'master' of github.com:OffchainLabs/nitro into jh/sync

ImJeremyHe committed:
Fix tests

ImJeremyHe committed:
Update contracts

ImJeremyHe committed:
Fix CI

zacshowa committed:
Changes to `handleEspressoPreConditions` in `espresso_validation.go` to fix incorrect logic, and give more granular control to where and how the generated panic is handled in the validators STF via returned closures.

zacshowa committed:
return nil when we need a no-op panic handler

zacshowa committed:
Fix lint errors

zacshowa committed:
Merge pull request #141 from zacshowa/update-validator

Refactor and expand validator preconditions to espresso STF logic

ImJeremyHe committed:
Merge pull request #142 from EspressoSystems/jh/dev-node

Update the nitro testnode

ImJeremyHe committed:
Update contracts

ImJeremyHe committed:
Fix the validation

ImJeremyHe committed:
Remove the switch test

ImJeremyHe committed:
Fix CI

ImJeremyHe committed:
Merge pull request #143 from EspressoSystems/jh/sync

Sync upstream 20240709

ImJeremyHe committed:
Remove postgres in espresso e2e test

ImJeremyHe committed:
Merge pull request #148 from EspressoSystems/jh/e2e

Remove postgres in espresso e2e test

sveitser committed:
Merge pull request #152 from Sneh1999/sync-3.1.0

Sync 3.1.0

sveitser committed:
Codeowners: add Alysia, Zach, Sneh (#154)

For #146

Sneh1999 committed:
Remove Dockerfile.espresso (#156)

* Remove Dockerfile.espresso

* address review comments

* add new line

sveitser committed:
Release 20240723 with 20240723-consensus

This should include all our changes for nitro v3.1.

- Check that the module root matches when downloading releases from
  github.

sveitser committed:
Merge pull request #160 from EspressoSystems/release-20240723

Release 20240723 with 20240723-consensus

sveitser committed:
Check all "not found" errors

Currently only the leveldb not found errors are handled. If a pebble not
found error is returned by the backend it is propagated, causing the
batchposter to fail to send transactions.

With this change all DB not found errors are handled the same way.

ImJeremyHe committed:
Merge pull request #175 from EspressoSystems/ma/fix-pebble-not-found-error

Check all "not found" errors

Sneh1999 committed:
Submit L2 Message to espresso network (#173)

* Submit L2 Message to espresso network

* add context timeout

* check for hotshot liveness and add a polling interval to see if transactions were included in an espresso block

* cleanup

* lint

* Draft the sovereign sequencer test

* Add espresso transaction queue

* Draft the espresso submission

* Set a new l2 message type for the sovereign sequencer transactions

* Attempt to write messages again

* Fix typos and add some comments

* Fix typo

* add header to jst and submit transaction to espresso

* Update go client for development

* Confirm the transaction inclusion in hotshot

* Fix the test

* Unblock the building blocks

* Update go client

* Fix the sovereign test

* add logs

* cleanup

* lint

* Updage go client

* not to fetch merkle proof if hotshot height is 0

* Cleanup

* Fix CI

* Add CI skip tests

* remove unwanted variable

* skipping some tests to make ci pass

* skipping some tests to make ci pass

---------

Co-authored-by: ImJeremyHe <yiliang.he@qq.com>

Sneh1999 committed:
Persistent storage for sovereign builder (#180)

* Persistence storage for sovereign builder

* add unit test

* add unit test to ci and add mutex

* Optional header in jst

* initialize pending transactions if not present

* remove unwanted test

* resolve bugs

* remove nil pointer panic

* cleanup

* revert to strings.contains

* fix isErrNotFound

---------

Co-authored-by: ImJeremyHe <yiliang.he@qq.com>

sveitser committed:
Save digest artifacts into own directory per run (#182)

* Save digest artifacts into own directory per run

Currently it seems that we have too many digests in the folder and I
think this breaks the assembly of the multiarch docker image.

With this change only digests for the same run are saved in the same
directory.

* Fail docker image merge on wrong number of digests
* Clean digests dir

sveitser committed:
Add dependabot rule to update nitro-testnode (#189)

ImJeremyHe committed:
Merge pull request #114 from EspressoSystems/jh/log

Clean the log with the log helper

sveitser committed:
Fix build (#191)

* Add missing dependency
* Use renamed variable, fix typo

sveitser committed:
dependabot: update contracts submodule (#193)

ImJeremyHe committed:
Pin the golint version in CI

ImJeremyHe committed:
Merge pull request #196 from EspressoSystems/jh/fix-go-ci

Pin the golint version in CI

Sneh1999 committed:
Fix pebble panic (#192)

* Fix pebble panic

* lint

sveitser committed:
Update some outdated help on flags (#203)

* Improve help texts for flags

Sneh1999 committed:
Update e2e tests to use Sovereign sequencer (#204)

* Update e2e tests to use Sovereign sequencer


Co-authored-by: ImJeremyHe <yiliang.he@qq.com>

ImJeremyHe committed:
Remove the global state for HotShot height (#210)

* Remove the global state

* Update the submodule

sveitser committed:
CI: use buildjet instead of self-hosted runner (#212)

* CI: use buildjet instead of self-hosted runner

* CI: use public runner / buildjet for docker build

This removes usage of our self-hosted github runners.

* CI: more disk space on public runner

* Check disk space at end of runs

* Try running CI tests from different directory

This may avoid failures due to IPC paths being too long

* Make workspace dir a symlink to /tmp/gha-work

* Fix redis URL

* Move redis tests to own workflow

sveitser committed:
CI: fix rust and go cache (#217)

* rust-cache: add more target dirs
* Don't set custom TMPDIR

This seems to be the culprit for the extremely long path names that
are too long for IPC paths.

zacshowa committed:
modify arbosstate.go to allow for upgrade to arbOS version 35 (#216)

* modify arbosstate.go to allow for upgrade to arbOS version 35

* Fix lint

* Modify the chain config during the ArbOS upgrade

---------

Co-authored-by: sveitser <sveitser@gmail.com>

sveitser committed:
CI: make more space (#218)

Sneh1999 committed:
Store transaction pos in batch poster (#219)

* Move submit transaction pos from tx_streamer to batch poster

Sneh1999 committed:
Remove espresso sequencer (#214)

* Remove espresso sequencer

* lint

* remove enable espresso

* commit go-ethereum

* fix solgen/gen.go

* fix build

* revert .gitmodules

* revert go.mod

* add enable espresso to validation check

* re-add enable espresso to tests

---------

Co-authored-by: Mathis <sveitser@gmail.com>

sveitser committed:
Update espresso-sequencer-go with new LC ABI (#230)

* Update espresso-sequencer-go with new LC ABI

Instead of using the block height from the first snapshot after the
height in the justification it uses the latest height in the LC
contract. Currently not 100% convinced about correctness but also
believe it should work because any newer tree still contains the header
of interest.

* Undo LC ABI changes

We're not comfortable with the previous change because making a proof
w.r.t to the newest finalized state may cause issues if the batch poster
is lagging behind. This is currently not easy to test so we decided to
revert back to the old flow.

* Use tag of espresso-sequencer-go

zacshowa committed:
Stop setting the chain config during ArbOS upgrade in favor of setting it manually during migration process. (#234)

Sneh1999 committed:
Use the tag before marketplace update (#238)

Sneh1999 committed:
Sign payload for sovereign builder (#181)

* Sign payload for sovereign builder

* save signed payload to db

* change parsing logic to include signature

* add some comments

* skip tests

* address comments

* add signature payload test

* test working

* fix logging and comments

* move placement of log

* fix ci

* fix ci

* only submit transaction if jst nil

* add log to debug ci

* use the latest instead of main for espresso-dev-node

* pin to legacy lc version

* pin to before marketplace version

* remove log

Sneh1999 committed:
Full node with espresso finality (#237)

* wip

* wip

* Add espresso finality node test

* fix ci

* fix ci

* pin to before marketplace version

* fix tests

* add comments

* address comments

* address comments

* address comments

* cleanup

* fix test

* fmt

* rerun ci

* ci skip tests

zacshowa committed:
Update to sequencer.go to remove `EnableEspressoSovereign` flag, and instead sequence based on hotshot liveness at l1 block height.

zacshowa committed:
Update node.go to remove reference to `EnableEspressoSovereign`

zacshowa committed:
update Espresso tests to reflect removed config variables, and add escape hatch test to e2e test.

zacshowa committed:
Add wait for hotshot liveness in sovereign sequencer test to ensure sovereign sequencer is enabled.

sveitser committed:
go fmt

zacshowa committed:
Add comment to explain changes needed in replay binary when panics are no longer an issue.

zacshowa committed:
Fix lint issues

zacshowa committed:
Merge branch 'escape-hatch-simple' of github.com:EspressoSystems/nitro-espresso-integration into escape-hatch-simple

zacshowa committed:
Modify construction of lightClientReader in sequencer.go to allow for nil `l1Reader`

zacshowa committed:
Update test configs to align with changes.

zacshowa committed:
Update comments according to feedback.

dependabot[bot] committed:
Bump nitro-testnode from `03095e0` to `884a63c` (#240)

Bumps [nitro-testnode](https://github.com/EspressoSystems/nitro-testnode) from `03095e0` to `884a63c`.
- [Commits](https://github.com/EspressoSystems/nitro-testnode/compare/03095e0bb682dc404526611f8f14d380477c5f04...884a63cd9895295ea06d864f7e2e1e7741529141)

---
updated-dependencies:
- dependency-name: nitro-testnode
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Mathis <sveitser@gmail.com>

zacshowa committed:
Revert removal of `EnableEspressoSovereign`

zacshowa committed:
Add additional check to ensure that if EnableEspressoSovereign is on we can't construct a sequencer with a nil lightClientReader

zacshowa committed:
Add `EnableEspressoSovereign` config flag to test configs.

zacshowa committed:
Merge pull request #243 from EspressoSystems/escape-hatch-simple

Escape Hatch

dependabot[bot] committed:
Bump contracts from `d226e9d` to `2ae1bbd` (#244)

Bumps [contracts](https://github.com/EspressoSystems/nitro-contracts) from `d226e9d` to `2ae1bbd`.
- [Release notes](https://github.com/EspressoSystems/nitro-contracts/releases)
- [Commits](https://github.com/EspressoSystems/nitro-contracts/compare/d226e9da617a52ebd2dccd0680715812f1071446...2ae1bbd0761d9d54a5aeb13dc26ef038caf87857)

---
updated-dependencies:
- dependency-name: contracts
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

sveitser committed:
Merge tag 'v3.2.1' into ma/sync-nitro-v3.2.1-ng

Arbitrum Nitro v3.2.1

WIP: currently running into same error as the CI when running `make
build-replay-env`:

    GOOS=wasip1 GOARCH=wasm go build -o target/machines/latest/replay.wasm ./cmd/replay/...
    target/bin/prover target/machines/latest/replay.wasm --generate-binaries target/machines/latest \
    -l target/machines/latest/forward.wasm -l target/machines/latest/soft-float.wasm -l target/machines/latest/wasi_stub.wasm -l target/machines/latest/host_io.wasm -l target/machines/latest/user_host.wasm -l target/machines/latest/arbcompress.wasm -l target/machines/latest/program_exec.wasm -l target/machines/latest/espresso_crypto.wasm
    Error: failed to validate WASM binary at "target/machines/latest/wasi_stub.wasm"

    Caused by:
    0: failed to validate target/machines/latest/wasi_stub.wasm
    1: reference-types not enabled: zero byte expected (at offset 0x18d8)

    Location:
        prover/src/binary.rs:329:10
    make: *** [Makefile:436: target/machines/latest/machine.wavm.br] Error 1

sveitser committed:
Pin rust to 1.81.0

See https://github.com/EspressoSystems/nitro-espresso-integration/issues/257

sveitser committed:
Set cargo config for espresso_crypto wasm build

This was recently added for the other wasm compilation steps.

sveitser committed:
Pin rust in E2E test

sveitser committed:
CI: remove deprecated rust toolchain action

sveitser committed:
Fix if statement

sveitser committed:
Regenerate data-poster external-signer test certs

(cherry picked from commit 8f961b0fc9f3ba5997c0d957759fb60a08552802)

sveitser committed:
Re-allow setting L1 stack config for tests

After the sync the code would always use the default values.

Don't think this is very clean but I couldn't think of something better
yet.

sveitser committed:
Fix trie node node found error during recording

I'm not exactly clear why we need to do this or if it's indeed the right
approach to solving the issue however it does seem to make it possible
to validate again.

Without it we get errors like

    INFO [10-30|12:24:29.127] DelayedSequencer: Sequenced              msgnum=1 startpos=1
    INFO [10-30|12:24:30.143] Reading message result remotely.         count=3
    WARN [10-30|12:24:32.143] prepareblocks failed to get state for block hdrNum=0 err="moved beyond genesis looking for state 0, genesis 0: missing trie node 147a5a93c0202b07e7d7e3c936a36a8d97913056045be3af23b95b0aa0b356c7 (path ) state 0x147a5a93c0202b07e7d7e3c936a36a8d97913056045be3af23b95b0aa0b356c7 is not available, not found"
    ERROR[10-30|12:24:32.143] Error while recording                    err="moved beyond genesis looking for state 0, genesis 0: missing trie node 147a5a93c0202b07e7d7e3c936a36a8d97913056045be3af23b95b0aa0b356c7 (path ) state 0x147a5a93c0202b07e7d7e3c936a36a8d97913056045be3af23b95b0aa0b356c7 is not available, not found" status=2

sveitser committed:
Fix the OSP test

- Enable 4844 blobs to avoid contract deployment error
- Fix address of OSP entry

sveitser committed:
Disable batch poster in Espresso finality node

We don't need this type of node to have a batch poster. Enabling it now
causes the test to fail because there's a check to ensure that redis is
used for the batch poster on the second node.

zacshowa committed:
Use ArbOS chain config as the cannonial chain config. (#252)

* Add checks to sequencer.go and transaction_streamer.go to utilize the chain config switch

* Update interface.go to export interface for `GetArbOSConfigAtHeight()`

* Add implementation of `GetArbOSConfigAtHeight()` for `ExecutionEngine`

* Add implementation of `GetArbOSConfigAtHeight()` for `ExecutionNode`

* Update `sequencer.go` and `transaction_streamer.go` to read from the `ArbOS` `chainConfig` as the cannonical `chainConfig` rather than the `chainConfig` held by the `Blockchain` struct.

* Add ArbOS chain config check in batch poster to stop adding espresso messages to the transaction streamers DB if EnableEspresso is false.

* Add missing error check in sequencer.go and nil check on arbos config in sequencer.go and batch_poster.go.

* Remove ineffectual assignment in sequencer.go

* Add test for new GetArbOSConfigAtHeight() method in ExecutionNode

This commit adds an integration test for the GetArbOSConfigAtHeight()
method on the Execution node that grabs the initial config, and
validates it, then updates the config and ensures that the new method
will fetch the updated config.

* restore nitro-testnode to proper commit

sveitser committed:
CI: increase parallelism

sveitser committed:
Merge remote-tracking branch 'origin/integration' into ma/sync-nitro-v3.2.1-ng

sveitser committed:
Fix linting

- Run `golang-lint run --fix`
- Check error after `waitForConfigUpdate`

sveitser committed:
Move finality node test to espresso tests

The race test detected a data race in

https://github.com/EspressoSystems/nitro-espresso-integration/actions/runs/11705919237/job/32601803846?pr=254

sveitser committed:
Update contracts submodule

sveitser committed:
Fix typo

ImJeremyHe committed:
Add wabt and cmake in wasm shell

ImJeremyHe committed:
Skip tests

ImJeremyHe committed:
Add tags in ci

ImJeremyHe committed:
Add ci skip tests

ImJeremyHe committed:
Add ci skip tests

ImJeremyHe committed:
Remove rerun-fails

This flag in gotestsum causes some bugs which multiple the test time.

https://github.com/gotestyourself/gotestsum/issues/417
https://github.com/gotestyourself/gotestsum/issues/423

sveitser committed:
CI: simplify

- Add scripts to debug more easily locally.
- Skip tests by substring match instead of exact match.
- Don't run coverage on CI, we never look at it.
- Move all Espresso tests to Espresso CI job.

sveitser committed:
Revert "Remove rerun-fails"

This reverts commit 6bef1c3c8707d0a9375566c910de3e37b67a4f8c.

sveitser committed:
Fix build with nix on macos (#296)

sveitser committed:
CI: install latest gotestsum

This enables the --hide-summary argument

sveitser committed:
Merge remote-tracking branch 'origin/integration' into ma/sync-nitro-v3.2.1-ng

sveitser committed:
Update submodule forks with merge commits

sveitser committed:
Merge pull request #254 from EspressoSystems/ma/sync-nitro-v3.2.1-ng

Sync nitro v3.2.1

dependabot[bot] committed:
Bump nitro-testnode from `07a25f1` to `77d1447` (#297)

Bumps [nitro-testnode](https://github.com/EspressoSystems/nitro-testnode) from `07a25f1` to `77d1447`.
- [Commits](https://github.com/EspressoSystems/nitro-testnode/compare/07a25f1df5fcfe2c65c0a772e9a60b701f2f1fbe...77d144747675f2a4c843bd068ac81175245ca813)

---
updated-dependencies:
- dependency-name: nitro-testnode
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

sveitser committed:
Add FFI bindings for Espresso validation (#294)

* Example call to espresso-crypto function stub in batch poster.

This commit includes a log line and example call to the espresso-crypto
function stubs in batch_poster.go

* Result of `make fmt`

Run `make fmt` to clean up formatting not apparent in editor

* Make current lib.rs c abi compatible

* result of make fmt

* Add C ABI ffi between rust crypto lib and espressocrypto go module.

This commit adds FFI support between exsisting Rust crypto helper code
and the exsisting espressocrypto go module in native.go These functions
will now serve as utility functions to use in the integration that will
handle ffi to the Rust code.

This commit also moves the espresso-crypto-helper code to the
espressocrypto directory as we no longer need it for the arbitrator.

* Build native.go successfully

* Update the espresso-crypto and add test for merkle proof verification

* update flake.nix to new go version and fix errors

* Current state of work

The current state of my work revolves around editing the Makefile to
include the new espressocrypto ffi functions in the build process.
Additionally, this commit restores some espressocrypto code to the
arbitrator in an attempt to get a docker image built via the test node.
I felt as though this would be easier for this PR as opposed to ripping
out all of the code from the arbitrator.

* Update the flake.lock

* Add the namespace proof test

* Current state of build system changes

This commit represents the current state of the build system after
testing today.
Currently things compile locally, but not when building an image from
the docker file. There are errors related to duplicate function
definitions from what seem to be the FFI lib that stylus exposes.

* Build nitro docker image sucessfully

- Modify makefile to have a file target. Since the build artifact of our
  static library is a file we should use this as the make target
  (instead of only having PHONY targets).

- Use rust 1.81 to build all shared rust libraries in docker.

Our crate requires a minimum version of 1.81. However the other
libraries like libstylus.a are currently compiled with 1.75 and this
seems to cause issues with duplicate symbol definitions during linking.

For example:

    8.718 go build -ldflags="-extldflags '-ldl' -X github.com/offchainlabs/nitro/cmd/util/confighelpers.version= -X github.com/offchainlabs/nitro/cmd/util/confighelpers.datetime= -X github.com/offchainlabs/nitro/cmd/util/confighelpers.modified=" -o target/bin/nitro "/workspace/cmd/nitro"
    42.74 # github.com/offchainlabs/nitro/cmd/nitro
    42.74 /usr/local/go/pkg/tool/linux_amd64/link: running clang failed: exit status 1
    42.74 /usr/bin/ld: /workspace/espressocrypto/lib/espresso-crypto-helper/target/release/libespresso_crypto_helper.a(std-1c4b19562077c20d.std.1f2242ed6435445e-cgu.0.rcgu.o): in function `__rdl_alloc':
    42.74 /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/alloc.rs:397: multiple definition of `__rdl_alloc'; /workspace/execution/gethexec/../../target/lib/libstylus.a(std-90f6ddbf82de36ec.std.409886f6357001f0-cgu.0.rcgu.o):/rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/alloc.rs:389: first defined here

Bump the rust version of the prover-header-builder to 1.81 and compile
the espresso crypto library in the same container.

* Fix LDFLAGS, remove unnecessary CFLAGS

* Fix some merge errors

- We need rust 1.81 to compile the espresso crypto lib
- Re-add makefile rules

* Fix merge errors

* Remove FFI call in batch poster to expidite merge.

* Update bf submodule in arbitrator to match upstream version (This was set to a different commit during my erroneous merge)

* Remove extra change in espressocrypto_wasm.go

* Fix error with temporary espressocryptowasm directory preventing CI from running

* fix error of missing function definitions in espressocryptowasm directory to allow CI to run

* Re-add nosec comment about gosec error that will be removed soon

* result of make fmt

---------

Co-authored-by: Zach Showalter <zacshowa@gmail.com>
Co-authored-by: ImJeremyHe <yiliang.he@qq.com>

ImJeremyHe committed:
Update the go client (#256)

Still using the legacy dev node

sveitser committed:
ci: fix `git ls-remote` error in docker build (#304)

* Use path context instead of git context

With the default git context the repository is re-cloned at the start of
the docker build and the .dockerignore file is ignored.

> Be careful because any file mutation in the steps that precede the
> build step will be ignored, including processing of the .dockerignore
> file since the context is based on the Git reference.

https://github.com/docker/build-push-action?tab=readme-ov-file#git-context

This leads to some `.git` worktree files showing up inside the
dockerbuild with triggers a bug in yarn.

See https://github.com/yarnpkg/yarn/issues/7537#issuecomment-1099584308
for more details about the yarn bug.

ImJeremyHe committed:
Use the latest espresso dev node (#303)

* Use the latest espresso dev node

- Fix an issue that may cause the validation reorg
- Waiting for L1 to be ready can avoid some potential bugs in dev node

zacshowa committed:
Add proof validation to batch poster (#301)

* Add espresso proof validation to batch poster

Adds calls to the espresso validation code to the batch poster to
validate the namespace proof and merkle proof in a TEE

* result of make fmt

* Fix lint errors

* Add better logging

* Fix batch poster after merge of upstream/integration

zacshowa committed:
Remove espresso replay binary modifications (#307)

* Remove espresso validation in the replay binary

* Remove espresso-crypto related dependencies in arbitrator

* remove espressocryptowasm

* Update Makefile and Dockerfile to remove references to unused espresso crypto wasm lib

* remove batch_poster.go.orig file

* Fixes from make fmt

* Remove espresso_ops_test

* Remove additional code related to espresso validation in the staker

* Remove changes on stateless block validator

* Undo the stopwaiter

Previous modifications on this was because we had to stop and start the seuquencers
to function the escape hatch. Right now we don't need that anymore. These changes are
from the upstream

* Rename test flag

---------

Co-authored-by: ImJeremyHe <yiliang.he@qq.com>

Sneh1999 committed:
Add attestation quote in batch poster (#306)

* Add attestation quote in batch poster

* write file

* update batch poster

* use read file

* update to userDataHash

* reduce file permissions

* nit

ImJeremyHe committed:
Submit multiple messages in one transaction (#308)

* Submit multiple messages in one transaction

* Remove the espresso_transaction_payload_signature_test.go

* Update go client

* Remove signature and data signer

* Remove arbitrator/espresso-crypto-helper

* Add size limit to hotshot transaction

---------

Co-authored-by: Sneh Koul <snehkoul1999@gmail.com>

sveitser committed:
Use upstream contracts submodule (#310)

We no longer need the customizations in nitro-contracts (except
potentially some deployment tools) so this commit resets it to the
before the celestia contracts were added

https://github.com/celestiaorg/nitro-contracts/commit/4f8c65c163a25296777d30813a10022b1d287b65

and makes the nitro code work with that version by removing the hotshot
/ light client address passed to the OSP.

ImJeremyHe committed:
Enhance readability

ImJeremyHe committed:
Remove the legacy code and fix the batch poster (#318)

* Remove the legacy code and fix the test

ImJeremyHe committed:
Add escape hatch

ImJeremyHe committed:
Lint

ImJeremyHe committed:
Lint and fix

ImJeremyHe commented on arbnode/transaction_streamer.go:
```diff
@@ -1653,6 +1692,96 @@ func (s *TransactionStreamer) submitEspressoTransactions(ctx context.Context, ig
 }
 
+func (s *TransactionStreamer) toggleEscapeHatch(ctx context.Context) error {
+	live, err := s.lightClientReader.IsHotShotLive(s.config().EspressoSwitchDelayThreshold)
+	if err != nil {
+		return err
+	}
+	// If hotshot is down, escape hatch is activated, the only thing is to check if hotshot is live again
+	if s.escapheHatchOn {
+		if live {
+			log.Info("HotShot is up, disabling the escape hatch")
+			s.escapheHatchOn = false
+		}
+		return nil
+	}
+
+	// If hotshot is up, escape hatch is disabled
+	// - check if escape hatch should be activated
+	// - check if the submitted transaction should be skipped from espresso verification
+	if !live {
+		log.Warn("enabling the escape hatch, hotshot is down")
+		s.escapheHatchOn = true
+	}
```
`escapeHatchOn` is only used to check if the batch poster should submit transactions to hotshot. If hotshot is `currently` down, then bypass it. If hotshot is `currently` up, then turn off the escape hatch. Don't we need a real time liveness to see if we should enable/disable the escape hatch?

ImJeremyHe replied:
I think using the L1 height in the message header is incorrect. HotShot blocks also have an L1 height. During the time when transaction approaching HotShot, the L1 height can be different.

Sneh1999 replied:
@zacshowa 


One thing to note here:
```
Say we have a message with l1 height x where hotshot was live based on the data stored in the light client
```

This doesnt apply anymore because earlier we had changes in the sequencer code, so we decided if we want to create an espresso message or not based on whether hotshot was live. But now, all we care about is that hotshot is able to include the l2Messages before it gets posted to L1 in real time. So I think Jeremy's current code behaviour seems right but let me know if you think otherwise!

Sneh1999 commented on arbnode/transaction_streamer.go:
```diff
@@ -1664,8 +1793,18 @@ func (s *TransactionStreamer) espressoSwitch(ctx context.Context, ignored struct
 	}
-	if config.ArbitrumChainParams.EnableEspresso {
-		return s.submitEspressoTransactions(ctx, ignored)
+	if config.ArbitrumChainParams.EnableEspresso && s.config().SovereignSequencerEnabled {
```
Should we delete `SovereignSequencerEnabled` here?

ImJeremyHe replied:
For finality nodes, they shouldn't check the escape hatch, so I use this condition.

Sneh1999 replied:
Can you expand further on this? Our code will only run inside the batch poster, so we wanted to check the escape hatch condition. 

Which finality nodes are we referring to here?

ImJeremyHe replied:
https://github.com/EspressoSystems/nitro-espresso-integration/blob/3248f9621ee418c0f3daf73a8706d66da5e3084c/system_tests/espresso_finality_node_test.go#L31

These code are in the tx streamer. So finality node in this test will also can run this function. I do think there should be a better way. I will try to refactor a bit

Sneh1999 replied:
Sorry that test is outdated, we will probably have to make changes to espresso finality node as well. So dont worry about the finality node right now. We will change it to not use `SovereignSequencerEnabled`

ImJeremyHe replied:
Add a comment on it. We can do more clean up after most of the stuff having being done

Sneh1999 reviewed (commented):
@ImJeremyHe 
Each batch poster configuration should allow the chain to choose between two behaviours: either waiting for Hotshot to go live before posting the batch, or, if Hotshot is not live, proceeding without checking batch consistency with Hotshot. 

We wanted to make this a config because then decision will shift to the chain. I am not sure if you were there but we discussed this in a standup. Let me know if this makes sense! Happy to answer any questions on this, if any

ImJeremyHe committed:
Config the escape hatch is on/off and fix submitting transactions

- If escape hatch is on, batch poster will bypass espresso validation
- If escape hatch is off, batch poster will wait for espresso being live

ImJeremyHe committed:
Lint

ImJeremyHe committed:
Config escape hatch

ImJeremyHe committed:
Add escape hatch test

ImJeremyHe committed:
Lint

ImJeremyHe committed:
Lint

Sneh1999 reviewed (approved):
LGTM @ImJeremyHe!

We should merge this PR in `celestia-integration` branch instead of the integration branch
